### ISSUE TITLE
Option to output lint result as JSON

### ISSUE THREAD

**Issue by nolanbconaway** [alanmcruickshank/sqlfluff#100]
I was toying around with creating a vscode linter extension for sqlfluff. The process entails wrapping a `sqlfluff lint` subprocess and parsing its output into the structure expected by vscode's linting API. 

It wouldn't be _very_ difficult to process the stdout as a string, but this reveals a possible need for a `--json` option so that the linting results can be machine readable in general. I am not sure how easy this would be to accomplish but it possible it might be very easy and a quick win.

Happy to work on this if all agree it is a good idea!

**Comment by barrywhart** [comment]
I think the internals of this will likely be pretty similar to the new `src/sqlfluff/diff_quality_plugin.py` module in PR #90. That PR implements a plugin for the `diff-quality` linting tool in the Python `diff_cover` package.

**Comment by nolanbconaway** [comment]
👍 interesting, so maybe this should hold for a sec while we get that merged in?

**Comment by alanmcruickshank** [comment]
I agree with both of you

@nolanbconaway - do you have a vision of what the structure of the json output would be? I'd love to agree that here before we get too far down the line. I've got a yaml output from the parsing engine which I've recently added and I'd love all the different output formats to be roughly in line with each other.

Great idea 👍

**Comment by nolanbconaway** [comment]
I saw the parser yaml option and we might just be able to convert that to json! In that line of thinking there is no _a priori_ reason why the output ought to be json as opposed to some other format such as yaml.

I haven't thought much about what the structure of the json might be, aside from handling my very narrow use case 😄 . Uncontroversially, I think the output should be informationally identical to the current human-readable output. In that respect (thinking as i type here...), this seems to be the hierarchical structure:

```json
{
    "file_path_or_stdin": {
        "errors": [
            {
                "line": 1, 
                "position": 2, 
                "id": "L006", 
                "message": "Operators should be preceded by a space."
            },
            {
                "line": 2, 
                "position": 2, 
                "id": "L006", 
                "message": "Operators should be preceded by a space."
            }
        ]
    }
}
```

Success/Fail flags could be added of course!

**Comment by barrywhart** [comment]
I don't necessarily think this has to wait until the other PR is merged; it's more that this one will probably be similar in some ways.

**Comment by nolanbconaway** [comment]
I did some digging into what that would require and hit a snag:

The linter prints errors on a per-file basis as it goes along, logging the output at [this line](https://github.com/alanmcruickshank/sqlfluff/blob/master/src/sqlfluff/linter.py#L552). But the JSON payload would need to be constructed across multiple files before being printed. So printing logic would need to be moved out of `Linter`, possibly to within the CLI handler itself or `LintedPath`? 

`LintedPath` already has a [`.violations()` method](https://github.com/alanmcruickshank/sqlfluff/blob/master/src/sqlfluff/linter.py#L237-L239) which gets us 90% of the way there, we would only need to serialize the violation object into a vanilla dictionary like

```python
dict(zip(('code', 'line_no', 'line_pos', 'description'), violation.get_info_tuple()))
```

and then everything should play nice with `json.dumps`.

I don't 100% understand how the linting and output printing work together, so please let me know if I am mistaken!

**Comment by nolanbconaway** [comment]
@alanmcruickshank @barrywhart resurrecting this issue!

Is there any general feedback on the JSON output structure proposed above?

```json
{
    "file/path": {
        "errors": [
            {
                "line": 1, 
                "position": 2, 
                "id": "L006", 
                "message": "Operators should be preceded by a space."
            },
            {
                "line": 2, 
                "position": 2, 
                "id": "L006", 
                "message": "Operators should be preceded by a space."
            }
        ]
    }
}
```

Similarly, I have not kept abreast of the changes in 0.3.x but would imagine printing is still incremental. 

We _could_ incrementally print JSON records at a per-file basis, in which each record is printed on a single line like:


```json
{"filepath": "path/to/file", "errors": [{"line": 1, "position": 2, "id": "L006", "message": "Operators should be preceded by a space."}]}
```

That's not 100% readable as-is since you still need to split newlines and loads the JSON. But! it would not require a revamp of the linter's printing functionality.

If we all agree on this JSON structure and printing approach I am happy to get cracking!

**Comment by barrywhart** [comment]
The output format looks good to me. One small suggestion: I suggest "path", not "file/path", because I think JSON requires keys to be text only (at least in actual JavaScript).

@alanmcruickshank's opinion is way more important than mine -- it is his project, I'm just a vocal contributor. 😄

**Comment by nolanbconaway** [comment]
Ah yeah very good point! Really we should have a `"filepath": "path/to/file"` key so that the full path is preserved.

Conveniently, this is what I proposed for the incremental printing scenario. So that's another reason to go that way.

**Comment by darrenhaken** [comment]
Is this closed because it's ready for use? 
@nolanbconaway did you end up writing a vscode plugin?

**Comment by nolanbconaway** [comment]
Yes, it is ready! The PR with that functionality was merged some time ago. Enjoy the JSON friend.
